cluster-autoscaler/estimator/binpacking_estimator.go

```diff
@@ -136,16 +137,16 @@ func (e *BinpackingNodeEstimator) tryToScheduleOnExistingNodes(
 		pod := pods[index]
 
 		// Check schedulability on all nodes created during simulation
-		nodeName, err := e.predicateChecker.FitsAnyNodeMatching(e.clusterSnapshot, pod, func(nodeInfo *framework.NodeInfo) bool {
+		nodeName, err := e.clusterSnapshot.SchedulePodOnAnyNodeMatching(pod, func(nodeInfo *framework.NodeInfo) bool {
 			return estimationState.newNodeNames[nodeInfo.Node().Name]
 		})
-		if err != nil {
+		if err != nil && err.Type() == clustersnapshot.NoNodesPassingPredicatesFoundError {
```

Comment:
My read on this function is that it is meant to short-circuit as soon as any pod in a list of Pending pods is encountered that can not be scheduled on an existing node in the cluster, and that it will return that pod, as well as any additional, not-yet-checked pods in the list to the caller. Why are we changing our approach here where only the "nominal" error (`NoNodesPassingPredicatesFoundError`) triggers that behavior? Currently any error response causes us to return that pod list, whereas now all other errors now trigger the return of a nil list.

The existing `FitsAnyNodeMatching` method seems to have comparable error paths as the new `SchedulePodOnAnyNodeMatching` method (though now strongly typed).

Reply:
The previous code didn't distinguish "unexpected" from "expected" errors in `FitsAnyNodeMatching` and `CheckPredicates`. I wanted to change that since it seems unsafe to just silently ignore them. But you're right, I forgot that `SchedulePodOnAnyNodeMatching` has two "expected" error types. Added `FailingPredicateError` to the check.

So now I think the only difference is that we error out if  `SchedulePodOnAnyNodeMatching` or `SchedulePod` returns an `SchedulingInternalError`, but that only happens in truly unexpected scenarios - if `ListNodeInfos` or `GetNodeInfo` fails. If they fail, adding the pod to the snapshot would fail anyway.

@BigDarkClown Could we get a third set of eyes on this? Do you think this would be better/easier to understand if we had the separate `PredicateResult` value?

Reply:
Yeah, I think in this case `PredicateResult` would be a bit more descriptive. I however still think we could make it work in the current form, just I would change the order of checks to be something like below. Then we would very explicitly say "if this is an internal error (unexpected), then return it, otherwise return the pod"

```
if err != nil && err.Type() == InternalErrorType {
  // Unexpected error.
  return nil, err
} else if err != nil {
  // Pod cannot be scheduled
  break
}
```

Reply:
That's indeed better, changed! I also added an explicit mention of checking against `SchedulingInternalError` to `ClusterSnapshot` method comments, and added comments at the callsites.

@jackfrancis WDYT?

Reply:
lol a week is a long time with this type of review detail, I had to revisit entirely

latest changes do not appear to change error flows, this lgtm

I'd like to push a follow-up to clean up some of this, though, look forward to that after this merges! :)